text/0035-opentelemetry-protocol.md

```diff
@@ -0,0 +1,425 @@
+
+## Table of Contents
+
+- [Motivation](#motivation)
+- [Protocol Details](#protocol-details)
+  - [Export Request and Response](#export-request-and-response)
+  - [OTLP over gRPC](#otlp-over-grpc)
+  - [Other Transports](#other-transports)
+- [Implementation Recommendations](#implementation-recommendations)
+  - [Multi-Destination Exporting](#multi-destination-exporting)
+- [Trade-offs and Mitigations](#trade-offs-and-mitigations)
+- [Future Versions and Interoperability](#future-versions-and-interoperability)
+- [Prior Art, Alternatives and Future Possibilities](#prior-art-alternatives-and-future-possibilities)
+- [Open Questions](#open-questions)
+- [Appendix A - Protocol Buffer Definitions](#appendix-a---protocol-buffer-definitions)
+- [Appendix B - Performance Benchmarks](#appendix-b---performance-benchmarks)
+- [Glossary](#glossary)
+- [Acknowledgements](#acknowledgements)
+- [Author's Address](#authors-address)
+
+
+## Motivation
+
+OTLP is a general-purpose telemetry data delivery protocol designed in the scope of OpenTelemetry project. It is an incremental improvement of OpenCensus protocol. Compared to OpenCensus protocol OTLP has the following improvements:
+
+- Ensures high reliability of data delivery and clear visibility when the data cannot be delivered. OTLP uses acknowledgements to implement reliable delivery.
+
+- It is friendly to Level 7 Load Balancers and allows them to correctly map imbalanced incoming traffic to a balanced outgoing traffic. This allows to efficiently operate large networks of nodes where telemetry data generation rates change over time.
+
+- Allows backpressure signalling from telemetry data destinations to sources. This is important for implementing reliable multi-hop telemetry data delivery all the way from the source to the destination via intermediate nodes, each having different processing capacity and thus requiring different data transfer rates.
+
+## Protocol Details
+
+OTLP defines the encoding of telemetry data and the protocol used to exchange data between the client and the server.
+
+This specification defines how OTLP is implemented over [gRPC](https://grpc.io/) and specifies corresponding [Protocol Buffers](https://developers.google.com/protocol-buffers/docs/overview) schema. Future extensions to OTLP may define implementations over other transports. For details of gRPC service definition see section [gRPC Transport](#grpc-transport).
+
+OTLP is a request/response style protocols: the clients send requests, the server replies with corresponding responses. This document defines one requests and response type: `Export`.
+
+### Export Request and Response
+
+After establishing the underlying transport the client starts sending telemetry data using `Export` requests.  The client continuously sends a sequence of `Export` requests to the server and expects to receive a response to each request:
+
+![Request-Response](images/otlp-request-response.png)
```

Comment:
in this picture do you assume that span is self contained or that it can have a reference to "resource" transmitted separately?

Reply:
It is self-contained. See protobuf definition below. Each request carries a Resource and a list of Spans. The ability to transmit Resource separately and optimize it away for re-transmissions is removed because with unary calls you are not guaranteed that if you have multiple server instances behind a load balancer sequentual calls will go to any particular server instance.